homeassistant/components/schedule/__init__.py

```diff
@@ -0,0 +1,305 @@
+        self._attr_icon = self._config.get(CONF_ICON)
+        self._attr_name = self._config[CONF_NAME]
+
+    @classmethod
+    def from_yaml(cls, config: ConfigType) -> Schedule:
+        """Return entity instance initialized from yaml storage."""
+        schedule = cls(config)
+        schedule.entity_id = f"{DOMAIN}.{config[CONF_ID]}"
+        schedule.editable = False
+        return schedule
+
+    async def async_update_config(self, config: ConfigType) -> None:
+        """Handle when the config is updated."""
+        self._config = STORAGE_SCHEMA(config)
+        self._attr_icon = config.get(CONF_ICON)
+        self._attr_name = config[CONF_NAME]
+        self._update()
+
+    @property
+    def capability_attributes(self) -> dict:
+        """Return the capability attributes."""
+        return {
+            ATTR_EDITABLE: self.editable,
+        }
+
+    @callback
+    def _clean_up_listener(self) -> None:
+        """Remove the update timer."""
+        if self._unsub_update is not None:
+            self._unsub_update()
+            self._unsub_update = None
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass."""
+        self.async_on_remove(self._clean_up_listener)
+        self._update()
+
+    @callback
+    def _update(self, _: datetime | None = None) -> None:
+        """Update the states of the schedule."""
+        now = dt_util.now()
+        todays_schedule = self._config.get(WEEKDAY_TO_CONF[now.weekday()], [])
+
+        # Determine current schedule state
+        self._attr_state = next(
+            (
+                STATE_ON
+                for time_range in todays_schedule
+                if time_range[CONF_FROM] <= now.time() <= time_range[CONF_TO]
+            ),
+            STATE_OFF,
+        )
+
+        # Find next event in the schedule
+        next_event = None
+        for day in range(7):
+            times = sorted(
+                itertools.chain(
+                    *[
+                        [time_range[CONF_FROM], time_range[CONF_TO]]
+                        for time_range in self._config.get(
+                            WEEKDAY_TO_CONF[(now.weekday() + day) % 7], []
+                        )
+                    ]
+                )
+            )
+            if next_event := next(
+                (
+                    possible_next_event
+                    for time in times
+                    if (
+                        possible_next_event := (
+                            datetime.combine(now.date(), time, tzinfo=now.tzinfo)
+                            + timedelta(days=day)
+                        )
+                    )
+                    > now
+                ),
+                None,
+            ):
+                break
+
+        self._attr_extra_state_attributes = {
+            day: self._config[day] for day in CONF_ALL_DAYS
+        } | {
+            ATTR_NEXT_EVENT: next_event,
```

Comment:
I wonder if `next_change` would better explain it's value.

Reply:
I've played with multiple. Also just `next`.
The thing with `next_change` is that we don't know it's a change (the next block could start at the time the other ends, thus the state remains active).

I went for `event` in this case, as it resembles an event on a calendar.

That said, I have no really strong feeling about it.